frontend/windows/handle_zip.go

```diff
@@ -0,0 +1,213 @@
+	gwclient "github.com/moby/buildkit/frontend/gateway/client"
+	"golang.org/x/exp/maps"
+	"golang.org/x/exp/slices"
+)
+
+const (
+	workerImgRef    = "mcr.microsoft.com/mirror/docker/library/ubuntu:jammy"
+	outputDir       = "/tmp/output"
+	buildScriptName = "_build.sh"
+)
+
+var (
+	varCacheAptMount = dalec.WithMountedAptCache("/var/cache/apt", "dalec-windows-var-cache-apt")
+	varLibAptMount   = dalec.WithMountedAptCache("/var/lib/apt", "dalec-windows-var-lib-apt")
+)
+
+func handleZip(ctx context.Context, client gwclient.Client, spec *dalec.Spec) (gwclient.Reference, *dalec.DockerImageSpec, error) {
+	sOpt, err := frontend.SourceOptFromClient(ctx, client)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	pg := dalec.ProgressGroup("Build windows container: " + spec.Name)
+	worker := workerImg(sOpt, pg)
+
+	bin, err := buildBinaries(spec, worker, sOpt)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to build binaries: %w", err)
+	}
+
+	st := getZipLLB(worker, spec.Name, bin)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	def, err := st.Marshal(ctx)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error marshalling llb: %w", err)
+	}
+
+	res, err := client.Solve(ctx, gwclient.SolveRequest{
+		Definition: def.ToPB(),
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	ref, err := res.SingleRef()
+	// Do not return a nil image, it may cause a panic
+	return ref, &dalec.DockerImageSpec{}, err
+}
+
+func specToSourcesLLB(spec *dalec.Spec, sOpt dalec.SourceOpts, opts ...llb.ConstraintsOpt) (map[string]llb.State, error) {
+	out := make(map[string]llb.State, len(spec.Sources))
+	for k, src := range spec.Sources {
+		isDir := dalec.SourceIsDir(src)
+
+		s := ""
+		switch {
+		case src.DockerImage != nil:
+			s = src.DockerImage.Ref
+		case src.Git != nil:
+			s = src.Git.URL
+		case src.HTTP != nil:
+			s = src.HTTP.URL
+		case src.Context != nil:
+			s = src.Context.Name
+		case src.Build != nil:
+			s = fmt.Sprintf("%v", src.Build.Source)
+		case src.Inline != nil:
+			s = "inline"
+		default:
+			return nil, fmt.Errorf("no non-nil source provided")
+		}
+
+		pg := dalec.ProgressGroup("Add spec source: " + k + " " + s)
+		st, err := src.AsState(k, sOpt, append(opts, pg)...)
+		if err != nil {
+			return nil, err
+		}
+
+		if isDir {
+			st = llb.Scratch().File(llb.Copy(st, "/", filepath.Join("/", k)))
+		}
+
+		out[k] = st
+	}
+
+	return out, nil
+}
+
+func mapToArraySortedByKeys[T any](m map[string]T) []T {
+	keys := dalec.SortMapKeys(m)
+	out := make([]T, 0, len(keys))
+	for _, k := range keys {
+		out = append(out, m[k])
+	}
+	return out
+}
+
+func buildBinaries(spec *dalec.Spec, worker llb.State, sOpt dalec.SourceOpts) (llb.State, error) {
+	sources, err := specToSourcesLLB(spec, sOpt)
+	if err != nil {
+		return llb.Scratch(), err
+	}
+
+	patchedMap := dalec.PatchSources(worker, spec, sources)
+
+	patched := mapToArraySortedByKeys(patchedMap)
+
+	combined := dalec.MergeAtPath(llb.Scratch(), patched, "/")
```

Comment:
Instead of merging states here, can we mount each one directly into the build?

Reply:
I couldn't do everything as a mount. This is because some are files which have the same target directory, whereas others are directories and can successfully be mounted to a subdirectory of "/build".

This means sources that are single files are being merged into one state, mounted at "/build", then the directory sources are mounted at subdirectories of "/build".

If there is a better way, please let me know and I'll modify this change.

for now, this is done.